aiohttp/client.py

```diff
@@ -611,34 +613,40 @@ async def _request(
                             resp.release()
 
                         try:
-                            parsed_url = URL(
+                            parsed_redirect_url = URL(
                                 r_url, encoded=not self._requote_redirect_url
                             )
-
                         except ValueError as e:
-                            raise InvalidURL(r_url) from e
+                            raise InvalidRedirectURL(r_url) from e
 
-                        scheme = parsed_url.scheme
+                        scheme = parsed_redirect_url.scheme
                         if scheme not in ("http", "https", ""):
                             resp.close()
-                            raise ValueError("Can redirect only to http or https")
+                            raise InvalidRedirectURL(
```

Comment:
@setla So I dug up an old discussion — #2507 — and have an improvement request.

I think that this specific case requires a subclass of `InvalidRedirectURL` for specificity, or perhaps a direct subclass of `ClientError`. Call it a `NonHttpUrlError` with a `yarl.URL` instance always exposed.

I suppose, this check should also be reproduced after line 407/409 for consistency — this way, the clients would be able to rely on `NonHttpUrlError` for both the initial request and redirects the same way.

The cases above and below are equivalent between each other but differ from the case here.

Such a change would allow clients to break out of the HTTP land and invoke other apps through `webbrowser.open()`, for example. It's something I've been wanting to have for half a decade and seems natural to do in this PR. Moreover, since it adds new exceptions, it would be weird to change them later if implemented separately. And it could facilitate #2335 later on.

Reply:
I see some conflict here. If `NonHttpURL` was subclass of `InvalidRedirectURL` I couldn't use it in lines 407/409 - it's not redirect. If it wasn't subclass of `InvalidRedirectURL` - it won't be catched if someone would like repair redirect url on his own (if raised) or will catch it if was raised in both of mentioned code places (again only one of them is redirect). This might require 2 new exceptions - `NonHttpRedirectURL` (subclass of `InvalidRedirectURL`) and `NonHttpURL` (subclass of `InvalidURL`) but in this case it's not scope of this issue and could be done in separate PR/issue. What do You think?

Reply:
@setla I think this should still happen in this PR so that there's no remaking of a feature with breaking changes later on.

To address your concern, I suggest using multiple inheritance as follows:
```mermaid

---
title: Exception hierarchy for invalid and non-HTTP URLs, direct or redirect
---

graph BT;

    base-client-error{{"
        ClientError
    "}};

    redirect-client-error["
        RedirectClientError
    "];

    invalid-url-client-error["
        InvalidUrlClientError
    "];

    non-http-url-client-error["
        NonHttpUrlClientError
    "];

    invalid-url-redirect-client-error[["
        InvalidUrlRedirectClientError
    "]];

    non-http-url-redirect-client-error[["
        NonHttpUrlRedirectClientError
    "]];

    subgraph "Base classes for the end-users to rely on";
        invalid-url-client-error
        redirect-client-error;
        non-http-url-client-error;
    end;

    invalid-url-client-error --> base-client-error;
    redirect-client-error --> base-client-error;
    non-http-url-client-error --> base-client-error;

    non-http-url-redirect-client-error --> redirect-client-error;
    non-http-url-redirect-client-error --> non-http-url-client-error;

    invalid-url-redirect-client-error --> invalid-url-client-error;
    invalid-url-redirect-client-error --> redirect-client-error;

```

With that, the end-users would have enough interfaces to catch specific error classes granularly.

Reply:
Isn't enough to have `NonHttpInvalidRedirectURL(InvalidRedirectURL, [NonHttpInvalidUrl])` and `NonHttpInvalidUrl(InvalidURL)`?  We also need to inherit `InvalidURL` for backward compatibility. Common class for redirect related errors can be added anytime (as one of parent class of redirect url errors) if needed but there is no usecase for it now

Reply:
There should be no thing called `NonHttpInvalidUrl` — non-HTTP URLs are valid, aiohttp just can't process them because they need to be process through something external to HTTP. So this naming is a misrepresentation of what's actually happening.

Reply:
As for having less exceptions in the hierarchy, I think having a full tree is well within the scope of this PR, in the context of #2507 at least.

Reply:
@setla here's an updated proposal:

```mermaid

---
title: Exception hierarchy for invalid and non-HTTP URLs, direct or redirect
---

graph BT;

    base-client-error{{"
        ClientError
    "}};

    base-value-error{{"
        ValueError
    "}};

    base-invalid-url-error{{"
        InvalidURL
    "}};

    redirect-client-error["
        RedirectClientError
    "];

    invalid-url-client-error["
        InvalidUrlClientError
    "];

    non-http-url-client-error["
        NonHttpUrlClientError
    "];

    invalid-url-redirect-client-error[["
        InvalidUrlRedirectClientError
    "]];

    non-http-url-redirect-client-error[["
        NonHttpUrlRedirectClientError
    "]];

    subgraph "Base classes for the end-users to rely on";
        invalid-url-client-error
        redirect-client-error;
        non-http-url-client-error;
    end;

    base-invalid-url-error --> base-value-error;
    base-invalid-url-error --> base-client-error;

    invalid-url-client-error --> base-invalid-url-error;
    redirect-client-error --> base-client-error;
    non-http-url-client-error --> base-client-error;

    non-http-url-redirect-client-error --> redirect-client-error;
    non-http-url-redirect-client-error --> non-http-url-client-error;

    invalid-url-redirect-client-error --> invalid-url-client-error;
    invalid-url-redirect-client-error --> redirect-client-error;

```

Reply:
Okay, just one more change - `NonHttpUrlClientError` needs to inherit `InvalidURL` to have the same init args. It seems reasonable to inherit `InvalidURL` anyway because this exception is related to some invalid url.

Reply:
> Okay, just one more change - `NonHttpUrlClientError` needs to inherit `InvalidURL` to have the same init args. It seems reasonable to inherit `InvalidURL` anyway because this exception is related to some invalid url.

No, those URLs aren't necessarily invalid. Just non-HTTP. It would be wrong to imply such a thing.